File: java/387_First_Unique_Character_in_a_String.java
/*
https://leetcode.com/problems/first-unique-character-in-a-string/
Given a string, find the first non-repeating character in it and return it's index. If it doesn't exist, return -1.

Count frequencies of each letter. Iterate over string again, return index of first letter with frequency of 1.
Time - O(n)
Space - O(1)
*/

public class Solution {
    public int firstUniqChar(String s) {

        int[] frequencies = new int[26];

        for (int i = 0; i < s.length(); ++i)
            ++frequencies[(int) s.charAt(i) - (int) 'a'];

        for (int i = 0; i < s.length(); ++i)
            if (frequencies[(int) s.charAt(i) - (int) 'a'] == 1)
                return i;
        return -1;
    }
}

File: python_1_to_1000/387_First_Unique_Character_in_a_String.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/first-unique-character-in-a-string/
# Given a string, find the first non-repeating character in it and return it's index. If it doesn't exist, return -1.

# Count frequency of each letter. Then iterate over s again, finding the index of the first letter with count of 1.
# Time - O(n)
# Space - O(1)

class Solution(object):
    def firstUniqChar(self, s):
        """
        :type s: str
        :rtype: int
        """
        counts = [0 for _ in range(26)]

        for c in s:
            counts[ord(c) - ord("a")] += 1

        for i, c in enumerate(s):
            if counts[ord(c) - ord("a")] == 1:
                return i

        return -1